### PR TITLE
New module: Add Icinga2 downtime module to allow users to schedule and remove downtime (monitoring/icinga2_downtime)

### PR DESCRIPTION
**Pull request by Shaita-KrZ**
##### SUMMARY
new module allowing users to schedule and remove downtime for icinga (able to work with regex)

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
icinga2_downtime.py

##### ANSIBLE VERSION
```
ansible 2.4.0
```

### CODE REVIEW

**Comment by Shaita-KrZ** [comment]
lib/ansible/modules/monitoring/icinga2_downtime.py:0:0: E203 requests import found, should use ansible.module_utils.urls instead.

Do I have to change that ? I mean I'm using "import requests" because it supports a fully restful API, and I don't really want to change all my code cause of this import.

thanks !

**Comment by sivel** [comment]
Yes, we have effectively blacklisted the use of the requests module.  To accept this we will need the module converted to using our module_utils package which requires no additional dependencies.

**Comment by Shaita-KrZ** [comment]
Alright so don't know if someone could help to use ansible.module_utils.urls instead of requests to do : 

r = requests.post(uri, data=json.dumps(payload), verify=False, headers=headers, auth=(icinga2_api_user, icinga2_api_password))

**Comment by sivel** [comment]
```
r = open_url(uri, data=json.dumps(payload), validate_certs=False, headers=headers, url_username=icinga2_api_user, url_password=icinga2_api_password)
```

You may also need `force_basic_auth=True` in there, I'm not sure how icinga works.

Also you will need to expose a validate_certs option through the module, defaulting to True, and allow the user to disable if needed.  It shouldn't default to false or be hard coded to false.

**Comment by Shaita-KrZ** [comment]
Alright "All checks have passed"

File: lib/ansible/modules/monitoring/icinga2_downtime.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+        required: false
+        default: Ansible
+        description:
+            - Name of the author. The author can be used to remove all the downtime scheduled by the C(author).
+    comment:
+        required: false
+        default: Downtime scheduled by Ansible
+        description:
+            - Comment text.
+'''
```
Comment (sivel) [inline lib/ansible/modules/monitoring/icinga2_downtime.py]: An option to control `validate_certs` should be included and default to false

File: lib/ansible/modules/monitoring/icinga2_downtime.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+  icinga2_downtime:
+    icinga2_url: http://localhost
+    icinga2_api_user: root
+    icinga2_api_password: icinga2
+    state: absent
+    author: Walid
+'''
+
+RETURN = '''
+#
```
Comment (sivel) [inline lib/ansible/modules/monitoring/icinga2_downtime.py]: All new modules should include actual RETURN data.

**Review by sivel — CHANGES_REQUESTED** [review]

**Comment by Shaita-KrZ** [comment]
Ok @sivel I made the change.

File: lib/ansible/modules/monitoring/icinga2_downtime.py [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+---
+module: icinga2_downtime
+version_added: "2.4"
+author: "Walid Shaita (@Shaita-KrZ)"
+short_description: Manage downtime for icinga2 using Icinga2 API.
+description:
+    - A module to manage downtime for icinga2 using Icinga2 API.
+    - This module is able to schedule or remove downtime for hosts or/and services (can be used with regex).
+requirements:
+    - requests
```
Comment (Lowess) [inline lib/ansible/modules/monitoring/icinga2_downtime.py]: Hi @Shaita-KrZ you should remove requests from the requirements

**Comment by Shaita-KrZ** [comment]
@sivel ?

**Comment by thmshmm** [comment]
Hi @Shaita-KrZ, nice module.

I have two changes you might want to add to your code. At line 432 there exists a problem on Python 2.6 where deltatime has no method total_seconds. You can fix it with these lines:

```
delta_time = datetime.timedelta(seconds=end_time_duration)
total_seconds = (delta_time.days * 24 * 3600) + delta_time.seconds + (delta_time.microseconds / 10**6)
end_time = start_time + total_seconds
```
And at line 283 you should check if the key "name" exists.

```
if message[0].has_key('name'):
    result['name'] = message[0]['name']
```

**Comment by yasa1987** [comment]
Is still something happening there ? This module would be useful for me...

**Comment by Lowess** [comment]
@yasa1987 @sivel Let me get in touch with @Shaita-KrZ who is in my team to make sure those changes you requested get updated so that we can hopefully have this merged soon !

Thanks !

**Pull request closed without merging**